runhouse/servers/http/http_utils.py

```diff
@@ -281,3 +313,205 @@ def handle_response(
+        raise HTTPException(
+            status_code=403, detail=f"Permission denied for file in path {path}"
+        )
+
+    except Exception as e:
+        raise HTTPException(
+            status_code=500, detail=f"Error reading file {path}: {str(e)}"
+        )
+
+    if binary_mode and isinstance(file_contents, bytes):
+        file_contents = file_contents.decode()
+
+    return Response(
+        data=file_contents,
+        output_type=OutputType.RESULT_SERIALIZED,
+        serialization=None,
+    )
+
+
+def folder_put(
+    path: Path,
+    overwrite: bool,
+    mode: str = None,
+    serialization: str = None,
+    contents: Dict[str, Any] = None,
+):
+    mode = mode or "wb"
+
+    if contents and not isinstance(contents, dict):
+        raise HTTPException(
+            status_code=422,
+            detail="`contents` argument must be a dict mapping filenames to file-like objects",
+        )
+
+    path.mkdir(parents=True, exist_ok=True)
+
+    if overwrite is False:
+        existing_files = {str(item.name) for item in path.iterdir()}
+        intersection = existing_files.intersection(set(contents.keys()))
+        if intersection:
+            raise HTTPException(
+                status_code=409,
+                detail=f"File(s) {intersection} already exist(s) at path: {path}",
+            )
+
+    for filename, file_obj in contents.items():
+        binary_mode = "b" in mode
+
+        if serialization:
+            file_obj = serialize_data(file_obj, serialization)
+
+        if binary_mode and not isinstance(file_obj, bytes):
+            file_obj = file_obj.encode()
+
+        file_path = path / filename
+        if not overwrite and file_path.exists():
+            raise HTTPException(
+                status_code=409, detail=f"File {file_path} already exists"
+            )
+
+        try:
+            with open(file_path, mode) as f:
+                f.write(file_obj)
+        except Exception as e:
+            HTTPException(status_code=500, detail=f"Failed to write file: {str(e)}")
+
+    return Response(output_type=OutputType.SUCCESS)
+
+
+def folder_ls(path: Path, full_paths: bool, sort: bool):
+    if not path.exists():
+        raise HTTPException(status_code=404, detail=f"Path {path} does not exist")
+
+    if not path.is_dir():
+        raise HTTPException(status_code=400, detail=f"Path {path} is not a directory")
+
+    paths = [p for p in path.iterdir()]
+
+    # Sort the paths by modification time if sort is True
+    if sort:
+        paths.sort(key=lambda p: p.stat().st_mtime, reverse=True)
+
+    # Convert paths to strings and format them based on full_paths
+    if full_paths:
+        files = [str(p.resolve()) for p in paths]
+    else:
+        files = [p.name for p in paths]
+
+    return Response(
+        data=files,
+        output_type=OutputType.RESULT_SERIALIZED,
+        serialization=None,
+    )
+
+
+def folder_rm(path: Path, contents: List[str], recursive: bool):
```

Comment:
a bit confused on contents and path both being options?

Reply:
this supports being able to specify specific folder contents to remove instead of just the entire folder itself, you think that's overkill?